pkg/crypto/rings/cache.go

```diff
@@ -0,0 +1,201 @@
+		log.Printf("DEBUG: Ring in cache, creating from points [%s]", appAddress)
+		ring, err = newRingFromPoints(points)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// return the ring
+	return ring, nil
+}
+
+// getRingForAppAddress returns the RingSinger used to sign relays. It does so by fetching
+// the latest information from the application module and creating the correct ring.
+// This method also caches the ring's public keys for future use.
+func (rc *ringCache) getRingForAppAddress(
+	ctx context.Context,
+	appAddress string,
+) (*ring.Ring, error) {
+	points, err := rc.getDelegatedPubKeysForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+	return newRingFromPoints(points)
+}
+
+// newRingFromPoints creates a new ring from points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for
+// the given application address, by querying the application module for it's
+// delegated pubkeys and converting them to points on the secp256k1 curve in
+// order to create the ring.
+func (rc *ringCache) getDelegatedPubKeysForAddress(
+	ctx context.Context,
+	appAddress string,
+) ([]ringtypes.Point, error) {
+	rc.ringPointsMu.Lock()
+	defer rc.ringPointsMu.Unlock()
+
+	// get the application's on chain state
+	req := &apptypes.QueryGetApplicationRequest{Address: appAddress}
+	res, err := rc.applicationQuerier.Application(ctx, req)
+	if err != nil {
+		return nil, ErrRingsAccountNotFound.Wrapf("address: %s [%v]", appAddress, err)
+	}
+
+	// create a slice of addresses for the ring
+	ringAddresses := make([]string, 0)
+	ringAddresses = append(ringAddresses, appAddress) // app address is index 0
+	if len(res.Application.DelegateeGatewayAddresses) < 1 {
+		// add app address twice to make the ring size of mininmum 2
+		// TODO_HACK: We are adding the appAddress twice because a ring
+		// signature requires AT LEAST two pubKeys. When the Application has
+		// not delegated to any gateways, we add the application's own address
+		// twice. This is a HACK and should be investigated as to what is the
+		// best approach to take in this situation.
+		ringAddresses = append(ringAddresses, appAddress)
+	} else if len(res.Application.DelegateeGatewayAddresses) > 0 {
+		// add the delegatee gateway addresses
+		ringAddresses = append(ringAddresses, res.Application.DelegateeGatewayAddresses...)
+	}
+
+	// get the points on the secp256k1 curve for the addresses
+	points, err := rc.addressesToPoints(ctx, ringAddresses)
+	if err != nil {
+		return nil, err
+	}
+
+	// update the cache overwriting the previous value
+	rc.ringPointsCache[appAddress] = points
+
+	// return the public key points on the secp256k1 curve
+	return points, nil
+}
+
+// addressesToPoints converts a slice of addresses to a slice of points on the
```

Comment:
Do you plan to add unit tests in this PR or a followup? If it's a follow up, please add a `TODO_TEST(@h5law)`

Reply:
No unit tests in this one: Potentially 2 (definietly 1) follow up PR for this pkg:

1. Listen for application module `EventDelegation` events
  - create the event and emit on delegation and undelegations
  - add events query client as dependency and use accordingly to invalidate address in cache
2. Potentially invalidate cache dependiing on a specified time period after inactivity
  - add block listener
  - invalidate inactive addresses in cache
    - track inactivity
    - 3 maps? points, access time, expirary time?

Reply:
First one sounds good. 

Second one might be a bit overkill since it'll be a lot of work to test cache features in a cache. I'd leave a note about it as an idea and move on.